### DISCUSSION TITLE
Nesting theme providers?

### DISCUSSION THREAD

**Discussion by moritzjacobs** [chakra-ui/chakra-ui#5312] [Q&A] [answered]
I hope the following explanation of our problem is understandable 😬 
Granted this is a very specific case and a rather unusual requirement, but here's what we would like to do:

We deploy a single Next.js build on multiple domains (a main brand, let's call it "ACME Corp." and a few sub brands, e.g. "ACME World Ltd." or "ACME Dynamite Co." etc), with different content by using Next's rewrite functionality. This works great, because 99% of our sites can share chakra theming and components and we can `extendTheme`, where they differ. There are only a few differences, mainly brand colors or typography things, but not all differences are that simple. Some components are completely different for certain sub brands. This all came together great so far because the way chakra theming works helps us to redcuce complicated code nicely 🎉 

```ts
// Something like this:

const useBrandTheme = (brandSlug) => {
  return useMemo(() => {
    if (brandSlug === "acmeWorld") {
      return extendTheme(baseTheme, acmeWorldTheme);
    }

    return extendTheme(baseTheme);
  }, [brandSlug]);
};

// …


function App() {
  const brandTheme = useBrandTheme("acme");

  return (
    <ChakraProvider theme={brandTheme}>
      …
    </ChakraProvider>
  );
}
```

However now we have a special requirement: on the main brand's home page (ACME) we want to display a section with teaser content from a sub brand (ACME World) and this section should then use component styles from that sub brand. 

Here's a sandbox outlining my original idea of how I thought it might work (nesting ChakraProviders): https://codesandbox.io/s/admiring-worker-7bbdz?file=/src/index.tsx

```ts
function App() {
  const brandTheme = useBrandTheme("acme");
  const subBrandTheme = useBrandTheme("acmeWorld");

  return (
    <ChakraProvider theme={brandTheme}>
      …

        <ChakraProvider theme={subBrandTheme}>
          …
        </ChakraProvider>
    </ChakraProvider>
  );
}

```

Alas, this does not work, because a nested theme provider seems to overwrite the classnames for the upper theme provider as well, so the whole page will now adhere to the sub brand's theme.

I also suspect even if we found a way to namespace the inner theme class names, it would create a lot of unnecessary duplicate CSS.

Any idea how to do this efficiently without doing complicated abstraction? Thanks for your time!

**Comment by primos63** [comment, marked as answer]
Take a look at my [CodeSandbox](https://codesandbox.io/s/chakra-ui-theme-extension-w5u2n?file=/src/components/ChakraNestedProvider/index.js). I think something like this is what you are looking for.

**Reply by primos63** [reply]
Did this answer your question?

**Reply by moritzjacobs** [reply]
Thanks for your input, I tried merging your nested provider in my example, but no luck yet: https://codesandbox.io/s/festive-hill-wz7fu?file=/src/index.tsx

**Reply by primos63** [reply]
This will fix it for you.
```jsx
    <ChakraProvider theme={brandTheme}>
      <VStack>
        <Heading>Main brand section</Heading>
        <Button>Foo Bar</Button>
        <Divider />
        {/* 
          You need a level which has an id for the cssVarsRoot. Box will be 
          the root for this section. Your cssVarRoot need to point to the id.
          So you use #foo not just foo.
          This becomes an app within an app. In my CSB, I have sections which
          helps to divide the mini-apps.
         */}
        <Box id="foo">
          <ChakraNestedProvider theme={subBrandTheme} cssVarsRoot="#foo">
            <VStack>
              <Heading>Sub brand section</Heading>
              <Button>Foo Bar</Button>
              <Divider />
            </VStack>
          </ChakraNestedProvider>
        </Box>

        <Heading>back to main brand</Heading>
        <Button>Foo Bar</Button>
      </VStack>
    </ChakraProvider>
  );
  ```

**Reply by moritzjacobs** [reply]
Yes, thank you very much!
